demo/mpfao/mpfao.c

```diff
@@ -202,6 +202,12 @@ int main(int argc, char **argv) {
   PetscBool perturb = PETSC_FALSE;
 
   ierr = TDyInit(argc, argv); CHKERRQ(ierr);
+  TDy  tdy;
+  ierr = TDyCreate(&tdy); CHKERRQ(ierr);
```

Comment:
This might be a silly question, but what do you think about merging these 4 functions, if they must be called in this order anyway? Something like (in this case):

```
ierr = TDyCreate(&tdy, RICHARDS, MPFA_O);
```
that would do the work of `TDyCreate`, `TDySetMode`, `TDySetDiscretizationMethod`, and `TDySetFromOptions` in that order, all at once? We could have utility functions for these that are hidden from view. But isn't it true that you either need all of this info in any case (or must assume some defaults if it's not provided)?

Reply:
By the way, I'm thinking that both supplied arguments here (the "mode" and the discretization) can serve as defaults that can be overridden by command line options. That gives it a conceptual consistency that makes merging them together feel unforced.

Reply:
Perhaps `TDyCreateDefault(comm, RICHARDS, MPFA_O, &tdy)` or `TDyCreateFromOptions(comm, &tdy)` and have it provide defaults and accept options.

PETSc used to have (and advocate using) a lot of convenience constructors (like `MatCreateSeqAIJ`), but they've caused more trouble than benefit. I don't know how sequential the setting of things needs to be -- I'd expect all the attributes should be possible to set in any order. Some "safer" APIs encode the "builder" process in the type system, but it's a bit cumbersome to do in C.

In any case, functions with lots of arguments are harder to understand (hard to remember what goes where) and more brittle in that any new argument breaks all existing use. In contrast, a sequence of `TDySetX` calls are easy to read and don't break code so long as defaults are available.

Reply:
We can merge, but we will still need those two separate functions on the library side.  Eventually, there will only one driver with those function calls.